frontend/pages/greetings/[name].tsx

```diff
@@ -1,38 +1,58 @@
-import React, { useState } from "react";
+import React, { useState, useEffect } from "react";
 import Link from "next/link";
-import fetch from "node-fetch";
+import axios from "axios";
 
 import { GetServerSideProps } from "next";
+import { getGreeting } from "../../lib/greetings";
 
 export const getServerSideProps: GetServerSideProps = async (context) => {
-  const res = await fetch(
-    `http://hello-world.hello-world/hello/${context?.params?.name}`
-  );
-
-  console.log("*****8888", typeof window);
-
-  const greeting = await res.text();
+  const nameProp = context?.params?.name || "";
+  const greeting = await getGreeting(nameProp);
 
   return {
     props: {
       greeting,
+      nameProp,
     },
   };
 };
 
-const GreetingForm = ({ greeting }: { greeting: object }) => {
+const GreetingForm = ({
+  greeting,
+  nameProp,
+}: {
+  greeting: object;
+  nameProp: string;
+}) => {
   const [inputValue, setInputValue] = useState("");
+  const [inputValueClient, setInputValueClient] = useState("");
+  const [axiosValue, setAxiosValue] = useState("");
+
+  const callAndSetGreeting = async (name: string) => {
+    try {
+      const res = await axios.get(`/hello/${name}`);
+
+      setAxiosValue(res.data);
+    } catch (e) {
+      console.error(e);
+    }
+  };
+
+  useEffect(() => {
+    callAndSetGreeting(nameProp);
+  }, []);
+
   return (
     <>
       <section>
+        <div>Server side rendering example</div>
         <div>Enter a name</div>
-
         <div>{greeting}</div>
         <input
           value={inputValue}
           onChange={(e) => setInputValue(e.target.value)}
         />
-
+        {/* shallow - Update the path of the current page without rerunning */}
```

Comment:
it's worth expanding on this with what the data flow looks like. I didn't believe how lightweight it was until you showed me the network tab.
```suggestion
        {/* shallow - Update the path of the current page and call getServerSideProps() on the server, then re-render on the client-side */}
```

It's also worth noting that given how lightweight the `<Link>`-based approach is, we should probably treat it as our default way of working.

Doing the fetch on the client side is probably a premature optimisation if it causes code duplication.

Reply:
yeah, next.js is really geared for server-side. client-side is so much more work.